### DISCUSSION TITLE
Common format for all derived types

### DISCUSSION THREAD

**Discussion by njansson** [ExtremeFLOW/neko#728] [Ideas]
(Based on the discussion in PR #720 )

Neko uses various ways of initialising objects, as shown below. Before the code grows too big, it would probably be a good idea to agree on a common format and make changes throughout fo 
r the next-next release (v0.6)

The various formats are shown below
```Fortran
type(apa_t) :: blargh

! F90 style
call apa_init(blargh, args)

! type bound procedures  
call blargh%init(args)

! type interface 
blargh = apa_t(args)
```
Overhead costs of type bound procedures aside, my biggest issue with enforcing all derived types to have `init()` methods is the somewhat reduced readability of the code's intent. Both the f90 and interface indicate what kind of type `myapa` is when calling the initialisation method. Any other thoughts?

**Comment by timofeymukha** [comment]
I would strike out the F90 option. Both `init` and the interface look fine, I think. Not sure I agree about the intent though, because to me the type interface is slightly more confusing, since it is an assignment.  But since all type names have a `_t` in the end, I think confusion with a regular function call is more or less impossible anyway. So I would just pick other init or interface and go for it!

**Reply by njansson** [reply]
The ide with the interface was to kind of mimic C++, `myapa = new Apa(args)` to at least see what the type is about. I did a test with only using `init()` and it quickly became rather unreadable.

**Reply by timofeymukha** [reply]
Let's go for the type interface, then unless others disagree strongly :-).

**Comment by timofeymukha** [comment]
@njansson 
https://github.com/fortrancon/workshop2020/blob/master/doc/tmu_constructors-and-patterns.pdf
Slide 5
